walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,156 @@
+    ArrayList<TimeRange> slots, boolean hasOptionalAttendees) {
+
+    // A copy of slots that stores new changes to slots without interferring
+    // with the for loop below.
+    ArrayList<TimeRange> curSlots = new ArrayList<>(slots);
+
+    for (Event event : events) {
+      TimeRange eventTime = event.getWhen();
+      slots = curSlots;
+
+      // Determine whether event's attendees overlaps with request's attendees.
+      if (hasOptionalAttendees && !hasAllAttendees(event, request)) {
+        continue;
+      }
+      if (!hasOptionalAttendees && !hasRequiredAttendees(event, request)) {
+        continue;
+      }
+
+      for (TimeRange slot : slots) {
+        TimeRange before = TimeRange.fromStartEnd(slot.start(), eventTime.start(), false);
+        TimeRange after = TimeRange.fromStartEnd(eventTime.end(), slot.end() - 1, true);
+
+        if (slot.contains(eventTime)) {
+          removeFromMiddle(curSlots, request.getDuration(), before, after, slot);
+          break;
+        } else if (slot.contains(eventTime.end())) {
+          removeFromEnd(curSlots, request.getDuration(), after, slot);
+          break;
+        } else if (slot.contains(eventTime.start())) {
+          removeFromStart(curSlots, request.getDuration(), before, slot);
+          break;
+        }
+      }
+    }
+    return curSlots;
+  }
+
+  /*
+   * Replaces slot with two new slots, with a gap in the middle corresponding
+   * to the space that the event takes (and thus the meeting cannot take place in).
+   */
+  private void removeFromMiddle(ArrayList<TimeRange> curSlots, long reqDuration,
+    TimeRange before, TimeRange after, TimeRange slot) {
+    if (before.duration() >= reqDuration) {
+      curSlots.add(before);
+    }
+    if (after.duration() >= reqDuration) {
+      curSlots.add(after);
+    }
+    curSlots.remove(slot);
+  }
+
+  /*
+   * Removes everything in the slot before TimeRange after, corresponding to the space
+   * that the event takes (and thus the meeting cnanot take place in).
+   */
+  private void removeFromEnd(ArrayList<TimeRange> curSlots, long reqDuration,
+    TimeRange after, TimeRange slot) {
+    if (after.duration() >= reqDuration) {
+      curSlots.add(after);
+    }
+    curSlots.remove(slot);
+  }
+
+  /*
+   * Removes everything in the slot after TimeRange before, corresponding to the space
+   * that the event takes (and thus the meeting cnanot take place in).
+   */
+  private void removeFromStart(ArrayList<TimeRange> curSlots, long reqDuration,
+    TimeRange before, TimeRange slot) {
+    if (before.duration() >= reqDuration) {
+      curSlots.add(before);
+    }
+    curSlots.remove(slot);
+  }
+
+  /*
+   * Return whether attendees for event overlap with required attendees for request
+   * (i.e. at least one attendee for request is an attendee for event).
+   */
+  private boolean hasRequiredAttendees(Event event, MeetingRequest request) {
```

Comment:
This method is only ever used with ! hasRequiredAttendees so perhaps it's more understandable if this determined if required attendees were *not* present during this event. 

I think it's especially confusing because hasRequiredAttendees makes it seems like this will return true if the event contains all the request's required attendees, which is not what this is doing.

Reply:
I changed the method names to `overlapWithRequiredAndOptionalAttendees` and `overlapWithRequiredAttendees`. I found that `overlapWithRequiredAttendees` is used without ! in a later function, so I ultimately kept both methods in the positive form in order to avoid using a double negative (i.e. `!noOverlapWithRequiredAttendees`).